### DISCUSSION TITLE
Foreign function interface in Val

### DISCUSSION THREAD

**Discussion by kcvinker** [hylo-lang/hylo#732] [Q&A] [answered]
Hi all,
I read about Val and got interested. I would like to know about the foreign function interface(FFI) in Val. How do I call a function like CreateWindowEx from a Windows DLL ? What about the support for 16 bit wchar string type ?

**Comment by kyouko-taiga** [comment, marked as answer]
Hello @kcvinker

Thanks for your interest!

We are only at a very early stage of the design for the FFI support. It is likely that we'll be able to declare and use FFIs as below in the source code and implement the necessary infrastructure to help the linker do its magic:
```swift
@ffi(c, CreateWindowEx)
fun create_window_ex(...) -> Handle

public fun main() {
  let handle = create_window_ex(...)
}
```

I don't see any specific reason why we couldn't support things like 16bit wide character strings, but I guess it might require some handwritten glue code to design all the types used by something like Windows API.

**Reply by ghost** [reply]
@kyouko-taiga @kcvinker 
I've never seen someone with my username before that's neat!

**Reply by kyouko-taiga** [reply]
Oops, apologies for the wrong mention.

**Comment by kcvinker** [comment]
@kyouko-taiga, Thank you for the explanation. And sorry for the delayed reply. Truly, this is the most easiest FFI declaration i have ever seen. I hope to learn Val soon
